src/chat.cpp

```diff
@@ -594,7 +607,30 @@ bool CHAT_Input( event_t *pEvent )
 			}
 			else if ( pEvent->data1 == '\t' )
 			{
-				g_ChatBuffer.TabComplete();
+				if (( g_ulChatMode == CHATMODE_PRIVATE_SEND ) && ( g_ulChatPlayer != MAXPLAYERS ))
+				{
+					int tempPlayer = g_ulChatPlayer;
+					int direction = ( pEvent->data3 & GKM_SHIFT ) ? -1 : 1;
+
+					do
+					{
+						tempPlayer += direction;
+
+						if ( tempPlayer < 0 )
+							tempPlayer = MAXPLAYERS - 1;
+						else if ( tempPlayer >= MAXPLAYERS )
+							tempPlayer = 0;
+
+						if ( tempPlayer == g_ulChatPlayer )
+							break;
+					}
+					while (( PLAYER_IsValidPlayer( tempPlayer ) == false ) || ( tempPlayer == consoleplayer ));
```

Comment:
Perhaps you want to skip bots? In that case, you need to check `SERVER_CountPlayers( false ) >= 2` instead of `SERVER_CalcNumPlayers() >= 2` in the comment above.

Reply:
Maybe so. Perhaps the client should only be allowed to use the private chat if the network state is `NETSTATE_CLIENT`?

Reply:
Good point. I agree that there is no need for private chat in single player or when watching demos.